Update Middleware documentation (#37992)
vercel/next.js: docs/advanced-features/middleware.md

---
description: Learn how to use Middleware to run code before a request is completed.
---

# Middleware

<details open>
<summary><b>Version History</b></summary>

| Version | Changes |

| --------- | ------------------------------------------------------------------------------------------ |

| `v12.2.0` | Middleware is stable |

| `v12.0.9` | Enforce absolute URLs in Edge Runtime ([PR](https://github.com/vercel/next.js/pull/33410)) |

| `v12.0.0` | Middleware (Beta) added |

</details>

Middleware allows you to run code before a request is completed, then based on the incoming request, you can modify the response by rewriting, redirecting, adding headers, or setting cookies.

> Middleware runs _before_ cached content, so you can personalize static files and pages. Common examples of Middleware would be authentication, A/B testing, localized pages, bot protection, and more.

Reviewer: ```suggestion
Middleware runs _before_ cached content, so you can personalize static files and pages. Common examples of Middleware would be authentication, A/B testing, localized pages, bot protection, and more.

The order middleware is applied is:

1. headers from `next.config.js`
2. redirects from `next.config.js`
3. middleware (rewrites, redirects, etc)
4. `beforeFiles` rewrites from `next.config.js`
5. filesystem routes (public/, _next/static/, static pages, etc)
6. `afterFiles` rewrites from `next.config.js`
7. dynamic routes (`/blog/[slug]`)
8. `fallback` rewrites from `next.config.js`
```
We should be sure to document the order middleware is applied in conjunction with custom routes in `next.config.js`

x-ref: https://github.com/vercel/next.js/issues/37797#issuecomment-1166311206

---
description: Learn how to use Middleware to run code before a request is completed.
---

# Middleware

<details open>
<summary><b>Version History</b></summary>

| Version | Changes |

| --------- | ------------------------------------------------------------------------------------------ |

| `v12.2.0` | Middleware is stable |

| `v12.0.9` | Enforce absolute URLs in Edge Runtime ([PR](https://github.com/vercel/next.js/pull/33410)) |

| `v12.0.0` | Middleware (Beta) added |

</details>

Middleware allows you to run code before a request is completed, then based on the incoming request, you can modify the response by rewriting, redirecting, adding headers, or setting cookies.

> Middleware runs _before_ cached content, so you can personalize static files and pages. Common examples of Middleware would be authentication, A/B testing, localized pages, bot protection, and more.

Reviewer 2: ```suggestion
Middleware runs _before_ cached content, so you can personalize static files and pages. Common use cases of Middleware would be authentication, A/B testing, localized pages, bot protection, and more.
```

Reviewer 3: I think examples and use cases are basically the same here

> Middleware runs _before_ cached content, so you can personalize static files and pages. Common examples of Middleware would be authentication, A/B testing, localized pages, bot protection, and more.

Reviewer 4: Are localized pages something we recommend middleware for, rather than [i18n routing](https://nextjs.org/docs/advanced-features/i18n-routing)? If so, I think we're gonna need to make it clear in what situations each one is appropriate to avoid confusion. If not, we should cut this part of the sentence.

Author: Addressed with this [commit](https://github.com/vercel/next.js/pull/37992/commits/2f42c0a5fcc38e3125536df4fdff1206fae053a7)

> **Note:** If you were using Middleware prior to `12.2`, please see the [upgrade guide](https://nextjs.org/docs/messages/middleware-upgrade-guide).

## Using Middleware

To begin using Middleware, follow the steps below:

1. Install the latest version of Next.js:

```bash

> npm install next@latest

Reviewer 4: This isn't necessary, latest is the default tag I think...

```suggestion
npm install next
```

```

> 2. Create a `middleware.ts` (or `.js`) file under your project root directory.

Reviewer 5: ```suggestion
2. Create a `middleware.ts` (or `.js`) file next to your `pages` folder.
```

Reviewer 4: I think this phrasing is maybe more confusing than "at the project root"

Reviewer 6: @jescalan This was addressing some feedback we'd got from users: https://vercel.slack.com/archives/C0289CGVAR2/p1656348479861419 Maybe we can be clearer in the language we're using though.

Author: Updated the wording in this [commit](https://github.com/vercel/next.js/pull/37992/commits/40806297a3112965814941945ae086f917761b56)

3. Export a middleware function from the `middleware.ts` file:

```typescript
// middleware.ts

> import { NextResponse } from 'next/server’

Reviewer 4: ```suggestion
import { NextResponse } from 'next/server'
```

```typescript
// middleware.ts

- [Authentication](https://github.com/vercel/examples/tree/main/edge-functions)
- [Bot protection](https://github.com/vercel/examples/tree/main/edge-functions)
- [Redirects and rewrites](https://github.com/vercel/examples/tree/main/edge-functions)
- [Handling unsupported browsers](https://github.com/vercel/examples/tree/main/edge-functions/user-agent-based-rendering)
- [Feature flags and A/B tests](https://github.com/vercel/examples/tree/main/edge-functions)
- [Advanced i18n routing requirements](https://github.com/vercel/examples/tree/main/edge-functions)
import { NextResponse } from 'next/server'
import type { NextRequest } from 'next/server'

## Execution Order
export function middleware(request: NextRequest) {
  if (request.nextUrl.pathname.startsWith('/about')) {
    return NextResponse.rewrite(new URL('/about-2', request.url))
  }

Middleware runs directly after `redirects` and `headers`, before the first filesystem lookup. This excludes `/_next` files.
  if (request.nextUrl.pathname.startsWith('/dashboard')) {
    return NextResponse.rewrite(new URL('/dashboard/user', request.url))
  }
}
```

## Deployment
## NextResponse

> Middleware uses a [strict runtime](/docs/api-reference/edge-runtime.md) that supports standard Web APIs like `fetch`. This works out of the box using `next start`, as well as on Edge platforms like Vercel, which use [Edge Functions](http://www.vercel.com/edge).

Reviewer 2: What is happening with edge functions? Is the relationship between middleware and edge functions changing? We should also address this since “edge functions” is a fairly popular term, especially since we’ve been using it interchangeably with middleware for a while

Is this repo still updated and relevant? If so, we might want to link to it: https://github.com/vercel/examples/tree/main/edge-functions

Reviewer 3: This is more of a Vercel thing than a Next.js thing 😄 Edge Functions and Edge Middleware are two separate things (previously poorly conflated).

> You **cannot change the response body** when using Middleware. If you need to change the response, consider using [Edge API Routes](/docs/api-routes/edge-api-routes.md)

Reviewer 3: ```suggestion
You **cannot change the response body** when using Middleware. If you need to send a custom response, consider using [Edge API Routes](/docs/api-routes/edge-api-routes.md)
```

> To produce a response from Middleware, you should `rewrite` to a route ([Page](/docs/basic-features/pages.md) or [Edge API Routes](/docs/api-routes/edge-api-routes.md)) that produces a response.

Reviewer 4: ```suggestion
To produce a response from Middleware, you should `rewrite` to a route ([Page](/docs/basic-features/pages.md) or [Edge API Route](/docs/api-routes/edge-api-routes.md)) that produces a response.
```

> </div>

Reviewer 4: This type of thing would be a prime use case for mdx 😉

- Set response cookies
- Set response headers

> To product a response from Middleware, you should `rewrite` to a route ([Page](/docs/basic-features/pages.md) or [Edge API Routes](/docs/api-routes/edge-api-routes.md)) that produces a response.

Reviewer 7: s/product/produce.

Maybe "To response with a response body…"

> To product a response from Middleware, you should `rewrite` to a route ([Page](/docs/basic-features/pages.md) or [Edge API Routes](/docs/api-routes/edge-api-routes.md)) that produces a response.

Reviewer 3: ```suggestion
To produce a response from Middleware, you should `rewrite` to a route ([Page](/docs/basic-features/pages.md) or [Edge API Routes](/docs/api-routes/edge-api-routes.md)) that produces a response.
```
